In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns",None)

**set_option():** Görüntülenecek varsayılan satır sayısını değiştirir ve iki argüman alır.

* **display.max_columns:** Görüntülenecek maksimum sütun sayısını görüntüler

In [2]:
import os
os.chdir(r"C:\Users\Excalibur\Desktop\VERİ BİLİMİ\KAGGLE\dataset")

In [3]:
dff = pd.read_csv("german_credit_data.csv")

Son aşamada veri ön işleme yapılmadan veriyi modele katıp farkı görmek için burada veriyi kopyalıyoruz.

In [4]:
df = dff.copy()

In [5]:
df.head()

Unnamed: 0  Age     Sex  Job Housing Saving accounts Checking account  \
0           0   67    male    2     own             NaN           little   
1           1   22  female    2     own          little         moderate   
2           2   49    male    1     own          little              NaN   
3           3   45    male    2    free          little           little   
4           4   53    male    2    free          little           little   

   Credit amount  Duration              Purpose  Risk  
0           1169         6             radio/TV  good  
1           5951        48             radio/TV   bad  
2           2096        12            education  good  
3           7882        42  furniture/equipment  good  
4           4870        24                  car   bad

* **Unnamed: 0** kolonu gereksizdir. Zaten datasetin atadığı bir indexleme söz konusu bu nedenle bu kolonu verisetinden kaldıralım.

In [6]:
df.drop("Unnamed: 0", axis=1, inplace=True)

In [7]:
df.head()

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0   67    male    2     own             NaN           little           1169   
1   22  female    2     own          little         moderate           5951   
2   49    male    1     own          little              NaN           2096   
3   45    male    2    free          little           little           7882   
4   53    male    2    free          little           little           4870   

   Duration              Purpose  Risk  
0         6             radio/TV  good  
1        48             radio/TV   bad  
2        12            education  good  
3        42  furniture/equipment  good  
4        24                  car   bad

In [11]:
for i in df.columns:
    print(i, ":",df[i].unique(),"\n\n")

Age : [67 22 49 45 53 35 61 28 25 24 60 32 44 31 48 26 36 39 42 34 63 27 30 57
 33 37 58 23 29 52 50 46 51 41 40 66 47 56 54 20 21 38 70 65 74 68 43 55
 64 75 19 62 59] 


Sex : ['male' 'female'] 


Job : [2 1 3 0] 


Housing : ['own' 'free' 'rent'] 


Saving accounts : [nan 'little' 'quite rich' 'rich' 'moderate'] 


Checking account : ['little' 'moderate' nan 'rich'] 


Credit amount : [ 1169  5951  2096  7882  4870  9055  2835  6948  3059  5234  1295  4308
  1567  1199  1403  1282  2424  8072 12579  3430  2134  2647  2241  1804
  2069  1374   426   409  2415  6836  1913  4020  5866  1264  1474  4746
  6110  2100  1225   458  2333  1158  6204  6187  6143  1393  2299  1352
  7228  2073  5965  1262  3378  2225   783  6468  9566  1961  6229  1391
  1537  1953 14421  3181  5190  2171  1007  1819  2394  8133   730  1164
  5954  1977  1526  3965  4771  9436  3832  5943  1213  1568  1755  2315
  1412 12612  2249  1108   618  1409   797  3617  1318 15945  2012  2622
  2337  7057  1469  2323 

## Aykırı Değerlerin Belirlenmesi

In [8]:
# df_numeric = df.select_dtypes(include = np.number)     ya da
"""df_numeric = df._get_numeric_data()    # sayısal sütunları al
df_numeric"""


"""for column in df_numeric:
    Q1, Q3 = df[column].quantile([0.25,0.75])
    IQR = Q3 - Q1
    lower_limit = Q1 - 1.5 * IQR
    upper_limit = Q3 + 1.5 * IQR
    
    no_outliers = df[df[column].between(lower_limit,upper_limit, inclusive = "both")][column]
    
    # inclusive = both : alt ve üst sınırı dahil eder.
    
    print(df[column].name)
    print("Alt Sınır:", lower_limit)
    print("Üst Sınır:", upper_limit)
    
    print("Aykırı değer sayısı:", df[column].shape[0] - no_outliers.shape[0])
    
    print("--------------------------------------------------")
    """

"""
for column in df_numeric:
    
    # Aykırı değerleri belirlediğimiz çeyreklikteki değerlere baskılıyalım.
    df.loc[df[column] < lower_limit, column] = lower_limit
    df.loc[df[column] > upper_limit, column] = upper_limit
"""

'\nfor column in df_numeric:\n    \n    # Aykırı değerleri belirlediğimiz çeyreklikteki değerlere baskılıyalım.\n    df.loc[df[column] < lower_limit, column] = lower_limit\n    df.loc[df[column] > upper_limit, column] = upper_limit\n'

In [12]:
q1 = df["Age"].quantile(0.25)
q3 = df["Age"].quantile(0.80)
iqr = q3 - q1
upper_limit = q3 + 1.5 * iqr
lower_limit = q1 - 1.5 * iqr

print("Üst sınır: ", upper_limit)
print("Alt sınır: ", lower_limit)

Üst sınır:  72.0
Alt sınır:  0.0


**Aykırı değerleri listeleyelim...**

In [13]:
df[(df["Age"] < lower_limit) | (df["Age"] > upper_limit)]

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
186   74  female    3    free          little         moderate           5129   
330   75    male    3    free          little           little           6615   
430   74    male    1     own          little              NaN           3448   
536   75  female    3     own             NaN           little           1374   
606   74    male    3     own          little              NaN           4526   
756   74    male    0     own          little             rich           1299   

     Duration   Purpose  Risk  
186         9       car   bad  
330        24       car  good  
430         5  business  good  
536         6       car  good  
606        24  business  good  
756         6       car  good

In [14]:
# Satırda veya sütunda aykırı değer var mı sorusunun cevabını döndürüyor.
df[(df["Age"] > upper_limit) | (df["Age"] < lower_limit)].any(axis=None)

True

**Aykırı değerleri belirlediğimiz çeyreklikteki değerlere baskılıyoruz.** Yani 0'dan küçük age değeri varsa onlara 0 atıyoruz aynı şekilde 72'den büyük bir değer varsa da onları 72'ye eşitliyoruz bu şekilde aykırı değerler baskılanıyor.

In [15]:
df.loc[df["Age"] < lower_limit, "Age"] = lower_limit

df.loc[df["Age"] > upper_limit, "Age"] = upper_limit

* Aykırı değer kontrolü yapıyoruz, baskılama işlemimiz sorunsuz gerçekleşmiş mi diye.
* False dönmesini bekliyoruz.

In [16]:
df[(df["Age"] > upper_limit) | (df["Age"] < lower_limit)].any(axis=None)

False

## Eksik değer sorunu ve null değerlerin doldurulması

In [17]:
df.isnull().sum() # değişkenlere göre boş değer sayısı

Age                   0
Sex                   0
Job                   0
Housing               0
Saving accounts     183
Checking account    394
Credit amount         0
Duration              0
Purpose               0
Risk                  0
dtype: int64

Veri setini yazdırarak aşağıdaki NaN değr bulunduran değişkenlerin yapısını gözlemleyelim.
* Saving accounts     183
* Checking account    394

In [18]:
df.head()[["Saving accounts","Checking account"]]

Saving accounts Checking account
0             NaN           little
1          little         moderate
2          little              NaN
3          little           little
4          little           little

NaN bulunduran değişkenler kategorik yapıdadır. Bu değişkenlerin eşsiz değerini gözlemleyelim.

In [19]:
df["Saving accounts"].value_counts()

little        603
moderate      103
quite rich     63
rich           48
Name: Saving accounts, dtype: int64

In [20]:
df["Checking account"].value_counts()

little      274
moderate    269
rich         63
Name: Checking account, dtype: int64

Object tipindeki değişkenlerin eşsiz sınıf sayısı 10'dan küçük olanların boş değerlerini mode'larıyla dolduralım.

* Yukarıdaki işlemlerden anlaşıldığı üzere iki değişkeninde en çok tekrar eden verisi "little" dır biz yine de ayrı ayrı tekrardan gözlemleyelim.

In [21]:
df["Saving accounts"].mode()[0]

'little'

In [22]:
df["Checking account"].mode()[0]

'little'

In [23]:
df = df.apply(lambda x: x.fillna(x.mode()[0]) if (x.dtype == "object" and len(x.unique()) <= 10) else x, axis = 0)

In [24]:
df.head()

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0   67    male    2     own          little           little           1169   
1   22  female    2     own          little         moderate           5951   
2   49    male    1     own          little           little           2096   
3   45    male    2    free          little           little           7882   
4   53    male    2    free          little           little           4870   

   Duration              Purpose  Risk  
0         6             radio/TV  good  
1        48             radio/TV   bad  
2        12            education  good  
3        42  furniture/equipment  good  
4        24                  car   bad

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Age               1000 non-null   int64 
 1   Sex               1000 non-null   object
 2   Job               1000 non-null   int64 
 3   Housing           1000 non-null   object
 4   Saving accounts   1000 non-null   object
 5   Checking account  1000 non-null   object
 6   Credit amount     1000 non-null   int64 
 7   Duration          1000 non-null   int64 
 8   Purpose           1000 non-null   object
 9   Risk              1000 non-null   object
dtypes: int64(4), object(6)
memory usage: 78.2+ KB


Sonuçtan elde ettiğimiz bilgiler:
* 1000 tane gözlem değeri var
* index aralığı 0-999
* Her değişkende yazan "1000 non-null" bize NaN (boş) değer olmadığını söylüyor.
* 4 int, 6 object tipinde değişkenimiz var

In [26]:
df["Age"].describe()

count    1000.000000
mean       35.532000
std        11.328959
min        19.000000
25%        27.000000
50%        33.000000
75%        42.000000
max        72.000000
Name: Age, dtype: float64

Standart sapma ile verilerin ne kadarının ortalamaya yakın olduğunu buluruz. Eğer standart sapma küçükse veriler ortalamaya yakın yerlerde dağılmışlardır. Bunun tersi olarak standart sapma büyükse veriler ortalamadan uzak yerlerde dağılmışlardır. Bütün değerler aynı olursa standart sapma sıfır olur.

* Age değişkenini belirlediğim quantile değerlerine göre kategorize ediyorum; genç,yetişkin,yaşlı olarak.

In [27]:
q25 = df["Age"].quantile(0.25)
q90 = df["Age"].quantile(0.90)

print("q25: {} \nq90: {}".format(q25,q90))

q25: 27.0 
q90: 52.0


In [28]:
df.loc[df["Age"] < q25, "new_age"] = "young"
df.loc[(df["Age"] >= q25) & (df["Age"] < q90), "new_age"] = "mature"
df.loc[df["Age"] >= q90, "new_age"] = "senior" 

In [29]:
df.head()

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0   67    male    2     own          little           little           1169   
1   22  female    2     own          little         moderate           5951   
2   49    male    1     own          little           little           2096   
3   45    male    2    free          little           little           7882   
4   53    male    2    free          little           little           4870   

   Duration              Purpose  Risk new_age  
0         6             radio/TV  good  senior  
1        48             radio/TV   bad   young  
2        12            education  good  mature  
3        42  furniture/equipment  good  mature  
4        24                  car   bad  senior

* İkili değerlere göre risk durumunu kontrol edelim.

In [30]:
def twice_col(df, col1, col2):
    
    c1 = df[col1].unique().tolist()
    c2 = df[col2].unique().tolist()
    
    for i in c1: # 1. kolonun benzersiz gözlemleri
        for j in range(len(c1)): # 1. kolondaki benzersizlerin sayısı
            for z in c2: # 2. kolonun benzersiz gözlemleri
                for t in range(len(c2)): # 2. kolondaki benzersizlerin sayısı
                    df.loc[(df[col1] == c1[j]) & (df[col2] == c2[t]), str(col1) + "_" + str(col2)] = str(c1[j]) + "_" + str(c2[t])


# datadaki kolondaki gözlem  eşit mi  benzersiz değişkenindeki gözleme,
# eğer eşitse önce kolon adlarını "_" ile birşetirip yeni sütun oluştur ardından oluşturduğumuz bu sütuna
# benzersizlerin ilgili değerlerini birleştirip yeni gözlem bilgisi ekle

In [31]:
twice_col(df, "new_age", "Sex") # kolon adı : new_age_Sex

In [32]:
twice_col(df, "Sex", "Housing") # kolon adı : Sex_Housing

In [33]:
twice_col(df, "new_age", "Housing") # kolon adı : new_age_Housing

In [34]:
twice_col(df, "Job", "Housing")  # kolon adı : Job_Housing

In [35]:
df.head()

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0   67    male    2     own          little           little           1169   
1   22  female    2     own          little         moderate           5951   
2   49    male    1     own          little           little           2096   
3   45    male    2    free          little           little           7882   
4   53    male    2    free          little           little           4870   

   Duration              Purpose  Risk new_age   new_age_Sex Sex_Housing  \
0         6             radio/TV  good  senior   senior_male    male_own   
1        48             radio/TV   bad   young  young_female  female_own   
2        12            education  good  mature   mature_male    male_own   
3        42  furniture/equipment  good  mature   mature_male   male_free   
4        24                  car   bad  senior   senior_male   male_free   

  new_age_Housing Job_Housing  
0      senior_own       2_own  
1       young_own       2_own  
2      mature_own       1_own  
3     mature_free      2_free  
4     senior_free      2_free

**Job değişkenine ait frekansları saydırıyorum, 4 unique değer olduğunu görüyorum.Herhangi bir işlem yapmamıza gerek yok.**

In [36]:
df.Job.value_counts()

2    630
1    200
3    148
0     22
Name: Job, dtype: int64

**Kredi ödeme sürelerinde fazla unique değer olduğundan model için 4 kategoriye ayıralım.**

In [37]:
df.Duration.unique()

array([ 6, 48, 12, 42, 24, 36, 30, 15,  9, 10,  7, 60, 18, 45, 11, 27,  8,
       54, 20, 14, 33, 21, 16,  4, 47, 13, 22, 39, 28,  5, 26, 72, 40],
      dtype=int64)

In [38]:
df.Duration.describe()

count    1000.000000
mean       20.903000
std        12.058814
min         4.000000
25%        12.000000
50%        18.000000
75%        24.000000
max        72.000000
Name: Duration, dtype: float64

In [39]:
df.loc[df.Duration < 18, "new_duration"] = "A"
df.loc[(df.Duration >= 18) & (df.Duration < 30), "new_duration"] = "B"
df.loc[(df.Duration >= 30) & (df.Duration < 48), "new_duration"] = "C"
df.loc[df.Duration >= 48, "new_duration"] = "D"

In [40]:
df.head()

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0   67    male    2     own          little           little           1169   
1   22  female    2     own          little         moderate           5951   
2   49    male    1     own          little           little           2096   
3   45    male    2    free          little           little           7882   
4   53    male    2    free          little           little           4870   

   Duration              Purpose  Risk new_age   new_age_Sex Sex_Housing  \
0         6             radio/TV  good  senior   senior_male    male_own   
1        48             radio/TV   bad   young  young_female  female_own   
2        12            education  good  mature   mature_male    male_own   
3        42  furniture/equipment  good  mature   mature_male   male_free   
4        24                  car   bad  senior   senior_male   male_free   

  new_age_Housing Job_Housing new_duration  
0      senior_own       2_own            A  
1       young_own       2_own            D  
2      mature_own       1_own            A  
3     mature_free      2_free            C  
4     senior_free      2_free            B

**Kredi ödeme miktarlarında da aynı şekilde birçok unique değer olduğundan kategorize edelim.**

In [41]:
df["Credit amount"].describe()

count     1000.000000
mean      3271.258000
std       2822.736876
min        250.000000
25%       1365.500000
50%       2319.500000
75%       3972.250000
max      18424.000000
Name: Credit amount, dtype: float64

In [42]:
q25 = df["Credit amount"].quantile(0.25)
q50 = df["Credit amount"].quantile(0.50)
q75 = df["Credit amount"].quantile(0.75)

In [43]:
df.loc[(df["Credit amount"] < q25), "new_Credit_amount"] = "A"
df.loc[(df["Credit amount"] >= q25) & (df["Duration"] < q50), "new_Credit_amount"] = "B"
df.loc[(df["Credit amount"] >= q50) & (df["Duration"] < q75), "new_Credit_amount"] = "C"
df.loc[(df["Credit amount"] >= q75),"new_Credit_amount"]= "D"

In [44]:
df.head()

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0   67    male    2     own          little           little           1169   
1   22  female    2     own          little         moderate           5951   
2   49    male    1     own          little           little           2096   
3   45    male    2    free          little           little           7882   
4   53    male    2    free          little           little           4870   

   Duration              Purpose  Risk new_age   new_age_Sex Sex_Housing  \
0         6             radio/TV  good  senior   senior_male    male_own   
1        48             radio/TV   bad   young  young_female  female_own   
2        12            education  good  mature   mature_male    male_own   
3        42  furniture/equipment  good  mature   mature_male   male_free   
4        24                  car   bad  senior   senior_male   male_free   

  new_age_Housing Job_Housing new_duration new_Credit_amount  
0      senior_own       2_own            A                 A  
1       young_own       2_own            D                 D  
2      mature_own       1_own            A                 B  
3     mature_free      2_free            C                 D  
4     senior_free      2_free            B                 D

In [45]:
df.Risk.unique()

array(['good', 'bad'], dtype=object)

Label encoder, kategorik değişkenleri numerik değişkenlere dönüştürmemizi sağlayan metottur.Risk de yukarıda görüldüğü gibi good ve bad olarak kategorize edilmiş bu yüzden 0 ve 1'e çevireceğiz.

In [46]:
from sklearn.preprocessing import LabelEncoder

In [47]:
le = LabelEncoder()
df.Risk = le.fit_transform(df.Risk)

In [48]:
df.head()  # Risk gözlemlerini 1 ve 0 numeric değerlerine döştürdük.

Age     Sex  Job Housing Saving accounts Checking account  Credit amount  \
0   67    male    2     own          little           little           1169   
1   22  female    2     own          little         moderate           5951   
2   49    male    1     own          little           little           2096   
3   45    male    2    free          little           little           7882   
4   53    male    2    free          little           little           4870   

   Duration              Purpose  Risk new_age   new_age_Sex Sex_Housing  \
0         6             radio/TV     1  senior   senior_male    male_own   
1        48             radio/TV     0   young  young_female  female_own   
2        12            education     1  mature   mature_male    male_own   
3        42  furniture/equipment     1  mature   mature_male   male_free   
4        24                  car     0  senior   senior_male   male_free   

  new_age_Housing Job_Housing new_duration new_Credit_amount  
0      senior_own       2_own            A                 A  
1       young_own       2_own            D                 D  
2      mature_own       1_own            A                 B  
3     mature_free      2_free            C                 D  
4     senior_free      2_free            B                 D

**Oluşturduğumuz kategorilere göre risk ortalamasına bakalım.**

In [51]:
df.groupby("Sex_Housing")["Risk"].mean()

Sex_Housing
female_free    0.421053
female_own     0.698980
female_rent    0.589474
male_free      0.629213
male_own       0.754352
male_rent      0.630952
Name: Risk, dtype: float64

In [53]:
df.groupby("new_age_Housing")["Risk"].mean()

new_age_Housing
mature_free    0.600000
mature_own     0.766932
mature_rent    0.628205
senior_free    0.571429
senior_own     0.791045
senior_rent    0.600000
young_free     0.600000
young_own      0.618056
young_rent     0.593407
Name: Risk, dtype: float64

In [54]:
def meann(df, b):
    a = {}
    for i in b:
        a[i] = df.groupby(i)["Risk"].mean()
    return a

In [55]:
b=['Sex_Housing','new_age_Housing','Job_Housing','new_age_Sex']

In [56]:
meann(df,b)

{'Sex_Housing': Sex_Housing
 female_free    0.421053
 female_own     0.698980
 female_rent    0.589474
 male_free      0.629213
 male_own       0.754352
 male_rent      0.630952
 Name: Risk, dtype: float64,
 'new_age_Housing': new_age_Housing
 mature_free    0.600000
 mature_own     0.766932
 mature_rent    0.628205
 senior_free    0.571429
 senior_own     0.791045
 senior_rent    0.600000
 young_free     0.600000
 young_own      0.618056
 young_rent     0.593407
 Name: Risk, dtype: float64,
 'Job_Housing': Job_Housing
 0_free    0.750000
 0_own     0.692308
 0_rent    0.600000
 1_free    0.375000
 1_own     0.753247
 1_rent    0.657895
 2_free    0.571429
 2_own     0.756637
 2_rent    0.573913
 3_free    0.666667
 3_own     0.638298
 3_rent    0.714286
 Name: Risk, dtype: float64,
 'new_age_Sex': new_age_Sex
 mature_female    0.682119
 mature_male      0.746032
 senior_female    0.794118
 senior_male      0.676056
 young_female     0.568000
 young_male       0.652174
 Name: Risk, dty

0 riskli olma durumu 1 olmama durumunu ifade ediyor.
* Örneğin female_free(kadın ve evsiz)'nin risk durumu yüksek mantıken kredisini de ödeyememesini bekleriz, female_own ise(kendi evi olan ve kadın) 'nın 1'e daha yakın yani kendi evi olan birinin kredisini ödemesini bekleriz

In [57]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Age                1000 non-null   int64 
 1   Sex                1000 non-null   object
 2   Job                1000 non-null   int64 
 3   Housing            1000 non-null   object
 4   Saving accounts    1000 non-null   object
 5   Checking account   1000 non-null   object
 6   Credit amount      1000 non-null   int64 
 7   Duration           1000 non-null   int64 
 8   Purpose            1000 non-null   object
 9   Risk               1000 non-null   int32 
 10  new_age            1000 non-null   object
 11  new_age_Sex        1000 non-null   object
 12  Sex_Housing        1000 non-null   object
 13  new_age_Housing    1000 non-null   object
 14  Job_Housing        1000 non-null   object
 15  new_duration       1000 non-null   object
 16  new_Credit_amount  1000 non-null   object
d

**Model için kategorik,numerik ve kategorik görünen fakat kardinal olan column'ları getiriyoruz.**

In [58]:
def grab_col_names(dataframe, cat_th=10, car_th=20):   # kardinal : bir araya gelere meclis oluşturan

    cat_cols = [col for col in dataframe.columns if dataframe[col].dtypes == "O"]
    
    num_but_cat = [col for col in dataframe.columns if dataframe[col].nunique() < cat_th and
                   dataframe[col].dtypes != "O"]

    cat_but_car = [col for col in dataframe.columns if dataframe[col].nunique() > car_th and
                   dataframe[col].dtypes == "O"]

    cat_cols = cat_cols + num_but_cat
    cat_cols = [col for col in cat_cols if col not in cat_but_car]

    num_cols = [col for col in dataframe.columns if dataframe[col].dtypes != "O"]
    num_cols = [col for col in num_cols if col not in num_but_cat]

    print(f"Gözlem Sayısı: {dataframe.shape[0]}")
    print(f"Değişken Sayısı: {dataframe.shape[1]}")
    print(f'Kateogorik kolon sayısı: {len(cat_cols)}')
    print(f'Numerik kolon sayısı: {len(num_cols)}')
    print(f'Kategorik görünen fakat kardinal olan kolon sayısı: {len(cat_but_car)}')
    print(f'Numeric görünen fakat kategorik olan kolon sayısı: {len(num_but_cat)}')

    # cat_cols + num_cols + cat_but_car = değişken sayısı.
    # num_but_cat cat_cols'un içerisinde zaten.
    # dolayısıyla tüm şu 3 liste ile tüm değişkenler seçilmiş olacaktır: cat_cols + num_cols + cat_but_car
    # num_but_cat sadece raporlama için verilmiştir.

    return cat_cols, cat_but_car, num_cols, num_but_cat

In [59]:
cat_cols, cat_but_car, num_cols, num_but_cat = grab_col_names(df)

Gözlem Sayısı: 1000
Değişken Sayısı: 17
Kateogorik kolon sayısı: 14
Numerik kolon sayısı: 3
Kategorik görünen fakat kardinal olan kolon sayısı: 0
Numeric görünen fakat kategorik olan kolon sayısı: 2


In [60]:
cat_cols

['Sex',
 'Housing',
 'Saving accounts',
 'Checking account',
 'Purpose',
 'new_age',
 'new_age_Sex',
 'Sex_Housing',
 'new_age_Housing',
 'Job_Housing',
 'new_duration',
 'new_Credit_amount',
 'Job',
 'Risk']

In [61]:
cat_but_car # Kategorik görünen fakat kardinal olan

[]

In [62]:
num_cols

['Age', 'Credit amount', 'Duration']

In [63]:
num_but_cat

['Job', 'Risk']

Burada ise tüm columnlarda gezilmesini, iki eşsiz sınıfa sahip olan ama int ve float olmayan columnları döndürmesini istiyoruz.Dönen columnlar kategorik olduğundan da label encoder ile numerik'e çeviriyoruz.

In [65]:
def label_encoder(df, col):
    return le.fit_transform(df[col])

binary_cols = [col for col in df.columns if df[col].dtype not in [int, float] and df[col].nunique() == 2]

for col in binary_cols:
    df[col] = label_encoder(df,col)

In [66]:
binary_cols

['Sex']

In [67]:
df = pd.DataFrame(df)

In [68]:
df.head()

Age  Sex  Job Housing Saving accounts Checking account  Credit amount  \
0   67    1    2     own          little           little           1169   
1   22    0    2     own          little         moderate           5951   
2   49    1    1     own          little           little           2096   
3   45    1    2    free          little           little           7882   
4   53    1    2    free          little           little           4870   

   Duration              Purpose  Risk new_age   new_age_Sex Sex_Housing  \
0         6             radio/TV     1  senior   senior_male    male_own   
1        48             radio/TV     0   young  young_female  female_own   
2        12            education     1  mature   mature_male    male_own   
3        42  furniture/equipment     1  mature   mature_male   male_free   
4        24                  car     0  senior   senior_male   male_free   

  new_age_Housing Job_Housing new_duration new_Credit_amount  
0      senior_own       2_own            A                 A  
1       young_own       2_own            D                 D  
2      mature_own       1_own            A                 B  
3     mature_free      2_free            C                 D  
4     senior_free      2_free            B                 D

Rare analyser fonksiyonu ile girilen dataframe,target ve kategorik columns parametrelerinin dataframe'deki sayısını, o column'daki değerlerin sayısının toplam dataframe sayısına oranını ve columnlara göre groupby yapıldığında target değişkeninin(Risk) ortalamasını bize döndürür.

In [69]:
def rare_analyser(dataframe, categorical_columns, target):
   
    rare_columns = [col for col in categorical_columns if (dataframe[col].value_counts() / len(dataframe)).any(axis=None)]

    for var in rare_columns:
        print(var, " : ", len(dataframe[var].value_counts()))

        print(pd.DataFrame({"COUNT": dataframe[var].value_counts(),
                            "RATIO": dataframe[var].value_counts() / len(dataframe),
                            "TARGET_MEAN": dataframe.groupby(var)[target].mean(),
                            "TARGET_MEDIAN": dataframe.groupby(var)[target].median()}),
              end="\n\n\n")

    print(len(rare_columns), " adet rare sınıfa sahip değişken var.")



In [70]:
df[["Sex","Job", "Risk"]] = df[["Sex","Job","Risk"]].astype("O")

In [71]:
cat_cols = df.select_dtypes(include=["object"]) # Kategorik kolonları bir değişkende tutuyoruz.

In [72]:
cat_cols

Sex Job Housing Saving accounts Checking account              Purpose  \
0     1   2     own          little           little             radio/TV   
1     0   2     own          little         moderate             radio/TV   
2     1   1     own          little           little            education   
3     1   2    free          little           little  furniture/equipment   
4     1   2    free          little           little                  car   
..   ..  ..     ...             ...              ...                  ...   
995   0   1     own          little           little  furniture/equipment   
996   1   3     own          little           little                  car   
997   1   2     own          little           little             radio/TV   
998   1   2    free          little           little             radio/TV   
999   1   2     own        moderate         moderate                  car   

    Risk new_age    new_age_Sex Sex_Housing new_age_Housing Job_Housing  \
0      1  senior    senior_male    male_own      senior_own       2_own   
1      0   young   young_female  female_own       young_own       2_own   
2      1  mature    mature_male    male_own      mature_own       1_own   
3      1  mature    mature_male   male_free     mature_free      2_free   
4      0  senior    senior_male   male_free     senior_free      2_free   
..   ...     ...            ...         ...             ...         ...   
995    1  mature  mature_female  female_own      mature_own       1_own   
996    1  mature    mature_male    male_own      mature_own       3_own   
997    1  mature    mature_male    male_own      mature_own       2_own   
998    0   young     young_male   male_free      young_free      2_free   
999    1  mature    mature_male    male_own      mature_own       2_own   

    new_duration new_Credit_amount  
0              A                 A  
1              D                 D  
2              A                 B  
3              C                 D  
4              B                 D  
..           ...               ...  
995            A                 B  
996            C                 C  
997            A                 A  
998            C                 B  
999            C                 D  

[1000 rows x 14 columns]

In [73]:
rare_analyser(df, cat_cols, "Risk")

#male:1 female:0 
#Bu dataframede male'den 310 gözlem female'den 690 gözlem vardır. Oranları 310/1000'den 0.31,
# 690/1000'den 0.69'dur.Ve erkeklerin risk durumu kadınlarınkinden daha yüksektir.
#Devamı da bu şekilde yorumlanabilir.

Sex  :  2
   COUNT  RATIO  TARGET_MEAN  TARGET_MEDIAN
0    310   0.31     0.648387            1.0
1    690   0.69     0.723188            1.0


Job  :  4
   COUNT  RATIO  TARGET_MEAN  TARGET_MEDIAN
0     22  0.022     0.681818            1.0
1    200  0.200     0.720000            1.0
2    630  0.630     0.704762            1.0
3    148  0.148     0.655405            1.0


Housing  :  3
      COUNT  RATIO  TARGET_MEAN  TARGET_MEDIAN
free    108  0.108     0.592593            1.0
own     713  0.713     0.739130            1.0
rent    179  0.179     0.608939            1.0


Saving accounts  :  4
            COUNT  RATIO  TARGET_MEAN  TARGET_MEDIAN
little        786  0.786     0.683206            1.0
moderate      103  0.103     0.669903            1.0
quite rich     63  0.063     0.825397            1.0
rich           48  0.048     0.875000            1.0


Checking account  :  3
          COUNT  RATIO  TARGET_MEAN  TARGET_MEDIAN
little      668  0.668     0.729042            1.0
modera

Rare encoder fonksiyonu kendisine parametre olarak girilen dataframe ve oranla ilgili işlem yapıyor, girilen oran parametresini alt sınır kabul edip o sınırın altındakileri rare olarak adlandırıyor ve adlandırılan birçok değişken olursa hepsi rare olduğundan oranları ve sayıları toplanıyor.Kısacası diğer olarak döndürülüyor. Dataframede az sayıda gözlemlendiğinden tek tek değerlendirip modeli de kendimizi de yormuyoruz.

In [74]:
def rare_encoder(dataframe, rare_perc):
   
    rare_columns = [col for col in dataframe.columns if (dataframe[col].value_counts() / len(dataframe) < rare_perc).any(axis=None)]

    for var in rare_columns:
        print(var, " : ", len(dataframe[var].value_counts()))

        print(pd.DataFrame({"COUNT": dataframe[var].value_counts(),
                            "RATIO": dataframe[var].value_counts() / len(dataframe)}))

    print(len(rare_columns), " adet rare sınıfa sahip değişken var.")


In [75]:
rare_encoder(df, 0.05)

Age  :  52
    COUNT  RATIO
27     51  0.051
26     50  0.050
23     48  0.048
24     44  0.044
28     43  0.043
25     41  0.041
30     40  0.040
35     40  0.040
36     39  0.039
31     38  0.038
29     37  0.037
32     34  0.034
33     33  0.033
34     32  0.032
37     29  0.029
22     27  0.027
40     25  0.025
38     24  0.024
42     22  0.022
39     21  0.021
46     18  0.018
47     17  0.017
43     17  0.017
41     17  0.017
44     17  0.017
45     15  0.015
21     14  0.014
20     14  0.014
49     14  0.014
50     12  0.012
48     12  0.012
54     10  0.010
52      9  0.009
57      9  0.009
51      8  0.008
63      8  0.008
55      8  0.008
53      7  0.007
61      7  0.007
72      6  0.006
60      6  0.006
64      5  0.005
66      5  0.005
65      5  0.005
58      5  0.005
67      3  0.003
68      3  0.003
56      3  0.003
59      3  0.003
19      2  0.002
62      2  0.002
70      1  0.001
Job  :  4
   COUNT  RATIO
2    630  0.630
1    200  0.200
3    148  0.148
0     22  0.02

In [76]:
rare_analyser(df, cat_cols, "Risk")

Sex  :  2
   COUNT  RATIO  TARGET_MEAN  TARGET_MEDIAN
0    310   0.31     0.648387            1.0
1    690   0.69     0.723188            1.0


Job  :  4
   COUNT  RATIO  TARGET_MEAN  TARGET_MEDIAN
0     22  0.022     0.681818            1.0
1    200  0.200     0.720000            1.0
2    630  0.630     0.704762            1.0
3    148  0.148     0.655405            1.0


Housing  :  3
      COUNT  RATIO  TARGET_MEAN  TARGET_MEDIAN
free    108  0.108     0.592593            1.0
own     713  0.713     0.739130            1.0
rent    179  0.179     0.608939            1.0


Saving accounts  :  4
            COUNT  RATIO  TARGET_MEAN  TARGET_MEDIAN
little        786  0.786     0.683206            1.0
moderate      103  0.103     0.669903            1.0
quite rich     63  0.063     0.825397            1.0
rich           48  0.048     0.875000            1.0


Checking account  :  3
          COUNT  RATIO  TARGET_MEAN  TARGET_MEDIAN
little      668  0.668     0.729042            1.0
modera

One hot encoder yine bir kategorik değişkenleri numerik'e dönüştürme yöntemidir.Kaç değişken döndürülecekse n-1 kadar column ile matrix oluşturulur.

In [77]:
ohe_cols = [col for col in df.columns if 12 >= df[col].nunique() > 2]
ohe_cols

['Job',
 'Housing',
 'Saving accounts',
 'Checking account',
 'Purpose',
 'new_age',
 'new_age_Sex',
 'Sex_Housing',
 'new_age_Housing',
 'Job_Housing',
 'new_duration',
 'new_Credit_amount']

In [78]:
#one_hot_encoder

def one_hot_encoder(dataframe, categorical_columns, nan_as_category=False):
    """
    Drop_first doğrusal modellerde yapılması gerekli

    Ağaç modellerde gerekli değil ama yapılabilir.

    dummy_na eksik değerlerden değişken türettirir.

    :param dataframe: İşlem yapılacak dataframe
    :param categorical_columns: One-Hot Encode uygulanacak kategorik değişken adları
    :param nan_as_category: NaN değişken oluştursun mu? True/False
    :return: One-Hot Encode yapılmış dataframe ve bu işlem sonrası oluşan yeni değişken adlarını döndürür.
    """
    original_columns = list(dataframe.columns)

    dataframe = pd.get_dummies(dataframe, columns=categorical_columns,
                               dummy_na=nan_as_category, drop_first=False)

    new_columns = [col for col in dataframe.columns if col not in original_columns]

    return dataframe, new_columns

# get_dummies(): Kategorik verileri kukla veya gösterge değişkenlerine dönüştürür.

In [79]:
df, one_hot_encodeds = one_hot_encoder(df, ohe_cols)

In [80]:
df.head()

Age Sex  Credit amount  Duration Risk  Job_0  Job_1  Job_2  Job_3  \
0   67   1           1169         6    1      0      0      1      0   
1   22   0           5951        48    0      0      0      1      0   
2   49   1           2096        12    1      0      1      0      0   
3   45   1           7882        42    1      0      0      1      0   
4   53   1           4870        24    0      0      0      1      0   

   Housing_free  Housing_own  Housing_rent  Saving accounts_little  \
0             0            1             0                       1   
1             0            1             0                       1   
2             0            1             0                       1   
3             1            0             0                       1   
4             1            0             0                       1   

   Saving accounts_moderate  Saving accounts_quite rich  Saving accounts_rich  \
0                         0                           0                     0   
1                         0                           0                     0   
2                         0                           0                     0   
3                         0                           0                     0   
4                         0                           0                     0   

   Checking account_little  Checking account_moderate  Checking account_rich  \
0                        1                          0                      0   
1                        0                          1                      0   
2                        1                          0                      0   
3                        1                          0                      0   
4                        1                          0                      0   

   Purpose_business  Purpose_car  Purpose_domestic appliances  \
0                 0            0                            0   
1                 0            0                            0   
2                 0            0                            0   
3                 0            0                            0   
4                 0            1                            0   

   Purpose_education  Purpose_furniture/equipment  Purpose_radio/TV  \
0                  0                            0                 1   
1                  0                            0                 1   
2                  1                            0                 0   
3                  0                            1                 0   
4                  0                            0                 0   

   Purpose_repairs  Purpose_vacation/others  new_age_mature  new_age_senior  \
0                0                        0               0               1   
1                0                        0               0               0   
2                0                        0               1               0   
3                0                        0               1               0   
4                0                        0               0               1   

   new_age_young  new_age_Sex_mature_female  new_age_Sex_mature_male  \
0              0                          0                        0   
1              1                          0                        0   
2              0                          0                        1   
3              0                          0                        1   
4              0                          0                        0   

   new_age_Sex_senior_female  new_age_Sex_senior_male  \
0                          0                        1   
1                          0                        0   
2                          0                        0   
3                          0                        0   
4                          0                        1   

   new_age_Sex_young_female  new_age_Sex_young_male  Sex_Housing_female_free  \
0                         0     

In [81]:
from sklearn.preprocessing import MinMaxScaler, LabelEncoder, StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Standartlaştırma

Şimdi ise numerik değerlerimizi standartlaştıracağız çünkü modele girerken bu işlemi yapmamız gerekir.Standartlaştırma, verileri belli bir sayı arasında oranlamaktır yani verileri aynı ölçeğe indirger.

In [82]:
num_cols = df.select_dtypes(["int","float"])

In [83]:
num_cols.columns

Index(['Age', 'Credit amount', 'Duration'], dtype='object')

In [84]:
scaler = StandardScaler()
df[num_cols.columns] = scaler.fit_transform(df[num_cols.columns])
df[num_cols.columns].head()

Age  Credit amount  Duration
0  2.779050      -0.745131 -1.236478
1 -1.195059       0.949817  2.248194
2  1.189407      -0.416562 -0.738668
3  0.836152       1.634247  1.750384
4  1.542661       0.566664  0.256953

In [85]:
df.head()

Age Sex  Credit amount  Duration Risk  Job_0  Job_1  Job_2  Job_3  \
0  2.779050   1      -0.745131 -1.236478    1      0      0      1      0   
1 -1.195059   0       0.949817  2.248194    0      0      0      1      0   
2  1.189407   1      -0.416562 -0.738668    1      0      1      0      0   
3  0.836152   1       1.634247  1.750384    1      0      0      1      0   
4  1.542661   1       0.566664  0.256953    0      0      0      1      0   

   Housing_free  Housing_own  Housing_rent  Saving accounts_little  \
0             0            1             0                       1   
1             0            1             0                       1   
2             0            1             0                       1   
3             1            0             0                       1   
4             1            0             0                       1   

   Saving accounts_moderate  Saving accounts_quite rich  Saving accounts_rich  \
0                         0                           0                     0   
1                         0                           0                     0   
2                         0                           0                     0   
3                         0                           0                     0   
4                         0                           0                     0   

   Checking account_little  Checking account_moderate  Checking account_rich  \
0                        1                          0                      0   
1                        0                          1                      0   
2                        1                          0                      0   
3                        1                          0                      0   
4                        1                          0                      0   

   Purpose_business  Purpose_car  Purpose_domestic appliances  \
0                 0            0                            0   
1                 0            0                            0   
2                 0            0                            0   
3                 0            0                            0   
4                 0            1                            0   

   Purpose_education  Purpose_furniture/equipment  Purpose_radio/TV  \
0                  0                            0                 1   
1                  0                            0                 1   
2                  1                            0                 0   
3                  0                            1                 0   
4                  0                            0                 0   

   Purpose_repairs  Purpose_vacation/others  new_age_mature  new_age_senior  \
0                0                        0               0               1   
1                0                        0               0               0   
2                0                        0               1               0   
3                0                        0               1               0   
4                0                        0               0               1   

   new_age_young  new_age_Sex_mature_female  new_age_Sex_mature_male  \
0              0                          0                        0   
1              1                          0                        0   
2              0                          0                        1   
3              0                          0                        1   
4              0                          0                        0   

   new_age_Sex_senior_female  new_age_Sex_senior_male  \
0                          0                        1   
1                          0                        0   
2                          0                        0   
3                          0                        0   
4                          0                        1   

   new_age_Sex_young_female  new_age_Sex_young_male  Sex_Housing_female_free  \
0      

## Modelleme

Veri ön işlemesini bitirdiğimiz veriyi modele katıyoruz.

In [98]:
X = df.drop("Risk", axis=1)
y = df["Risk"].astype("int")

# y object tiptedir, hata verdiği için int çevirdik. 

In [99]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=10)

In [100]:
rf_model = RandomForestClassifier(n_estimators=10, random_state=46)
rf_model.fit(X_train,y_train)

RandomForestClassifier(n_estimators=10, random_state=46)

In [101]:
y_pred = rf_model.predict(X_test)

accuracy_score(y_pred, y_test)

0.655

Modelin ön işlemesini yaptığımız parametrelere göre hedef değişkeni tahmin etme oranı %65.5 'miş gayet iyi.

Şimdi ise veri ön işlemenin önemini çok daha iyi anlayabilmek için veri ön işleme yapmadığımız yani verinin ilk okutulduğu halini modele katıyoruz.

In [102]:
dff.dropna(inplace=True) # Eksik değerleri kaldırdık.

In [103]:
dff = pd.get_dummies(dff, columns=["Sex"], drop_first=True)

# dff verisetindeki "Sex" kolonunu manipüle ediyoruz.
# drop_first: k kategorik seviyeden k-1 kukla almak için ilk seviyeyi kaldırın.

In [104]:
X = dff.drop(["Unnamed: 0","Risk","Housing","Saving accounts","Checking account","Purpose"], axis=1)
y = dff.Risk

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=10)

rf_model = RandomForestClassifier(random_state=46).fit(X_train,y_train)

y_pred = rf_model.predict(X_test)

accuracy_score(y_pred, y_test)

0.580952380952381

Görüldüğü üzere tahmin başarımız ön işleme yapılmadan %58.0 iken ön işleme yaptığımızda %65.5'e çıkmış. Kısacası veri ön işleme işlemi ne kadar iyi yapılırsa veriden o kadar anlamlı sonuçlar elde ederiz. 